aptos-move/framework/supra-framework/sources/automation_registry_state.move

```diff
@@ -0,0 +1,623 @@
+        gas_price_cap: u64,
+        registration_epoch: u64,
+        tx_hash: vector<u8>,
+    ) acquires AutomationRegistryState {
+        let registry_data = borrow_global_mut<AutomationRegistryState>(@supra_framework);
+        let registration_time = timestamp::now_seconds();
+
+        assert!(expiry_time > registration_time, EINVALID_EXPIRY_TIME);
+        assert!(gas_price_cap > 0, EINVALID_GAS_PRICE);
+        assert!(max_gas_amount > 0, EINVALID_MAX_GAS_AMOUNT);
+        assert!(vector::length(&tx_hash) == TXN_HASH_LENGTH, EINVALID_TXN_HASH);
+
+        let committed_gas = registry_data.gas_committed_for_next_epoch + max_gas_amount;
+        assert!(committed_gas < registry_data.automation_gas_limit, EGAS_AMOUNT_UPPER);
+        registry_data.gas_committed_for_next_epoch = committed_gas;
+        let task_index = registry_data.current_index;
+
+        let automation_task_metadata = AutomationTaskMetaData {
+            id: task_index,
+            owner: signer::address_of(owner),
+            payload_tx,
+            expiry_time,
+            max_gas_amount,
+            gas_price_cap,
+            state: PENDING,
+            registration_epoch,
+            registration_time,
+            tx_hash,
+        };
+
+        enumerable_map::add_value(&mut registry_data.tasks, task_index, automation_task_metadata);
+        registry_data.current_index = registry_data.current_index + 1;
+        event::emit(automation_task_metadata);
+    }
+
+    /// Cancel Automation task with specified id.
+    /// If the task was active its state is updated to be CANCELLED. Otherwise task is removed form the list.
+    /// Committed gas-limit is updated accordingly.
+    /// Only existing task can be cancled and only by task onwer.
+    public (friend) fun cancel_task(owner: &signer, id: u64): AutomationTaskMetaData acquires AutomationRegistryState {
+        let state = borrow_global_mut<AutomationRegistryState>(@supra_framework);
+        assert!(enumerable_map::contains(&state.tasks, id), EAUTOMATION_TASK_NOT_FOUND);
+
+        let automation_task_metadata = enumerable_map::get_value(&state.tasks, id);
+        assert!(automation_task_metadata.owner == signer::address_of(owner), EUNAUTHORIZED_TASK_OWNER);
+        assert!(automation_task_metadata.state != CANCELLED, EINVALID_CANCELLATION);
+        if (automation_task_metadata.state == PENDING) {
+            enumerable_map::remove_value(&mut state.tasks, id);
+        } else if (automation_task_metadata.state == ACTIVE) {
+            automation_task_metadata.state = CANCELLED;
+            enumerable_map::update_value(&mut state.tasks, id, automation_task_metadata);
+        };
+
+        // Adjust the gas committed for the next epoch by subtracting the gas amount of the cancelled task
+        state.gas_committed_for_next_epoch = state.gas_committed_for_next_epoch - automation_task_metadata.max_gas_amount;
+
+        event::emit(CanclledAutomationTask { id: automation_task_metadata.id });
```

Comment:
```suggestion
        event::emit(CancelledAutomationTask { id: automation_task_metadata.id });
```
Upon cancellation some event should also emit information about amount refunded

Reply:
I assume when coin transfer happens, there will be deposit event, wouldn;t that be enough or we want to have it part of cancellation event as well along with id?